File: Backend/manager_service/src/main/java/com/example/demo/services/TeamService.java
package com.example.demo.services;

import java.util.List;

import org.springframework.stereotype.Service;

import com.example.demo.entities.Team;
import com.example.demo.entities.User;
import com.example.demo.repositories.QueryRepository;
import com.example.demo.repositories.TeamRepository;
import com.example.demo.repositories.UserRepository;

import jakarta.transaction.Transactional;

@Service
public class TeamService {

    private final TeamRepository teamRepo;
    private final QueryRepository queryRepo;
    private final UserRepository userRepo;

    public TeamService(TeamRepository teamRepo, QueryRepository queryRepo, UserRepository userRepo) {
        this.teamRepo = teamRepo;
        this.queryRepo = queryRepo;
        this.userRepo = userRepo;
    }

    public Team addMember(Long pid, Long uid) {
        if (teamRepo.existsByUidAndPid(uid, pid)) {
            throw new RuntimeException("User already part of this project");
        }
        Team team = new Team();
        team.setPid(pid);
        team.setUid(uid);
        return teamRepo.save(team);
    }

    public Team createTeam(Long pid, Long domainId, Long tlUid) {
        if (teamRepo.existsByUidAndPid(tlUid, pid)) {
            throw new RuntimeException("Team Leader already assigned to this project");
        }
        Team team = new Team();
        team.setPid(pid);
        team.setUid(tlUid);
        return teamRepo.save(team);
    }

    public List<User> getTeamLeaders() {
        return userRepo.findByRole_Rname("Team Leader");
    }

    @Transactional
    public void removeMember(Long pid, Long uid) {
        Team team = teamRepo.findByUidAndPid(uid, pid)
                .orElseThrow(() -> new RuntimeException("Team member not found"));
        try {
            queryRepo.deleteQueriesByTeamId(team.getTeamId());
        } catch (Exception ignored) {}
        teamRepo.deleteByUidAndPid(uid, pid);
    }

    @Transactional
    public void deleteTeam(Long teamId) {
        Team team = teamRepo.findById(teamId)
                .orElseThrow(() -> new RuntimeException("Team not found"));
        try {
            queryRepo.deleteQueriesByTeamId(teamId);
        } catch (Exception ignored) {}
        teamRepo.deleteById(teamId);
    }

    public List<Team> getMembersByProject(Long pid) {
        return teamRepo.findByPid(pid);
    }
}

File: frontend/reactapp/src/services/teamService.js
import { managerAPI } from './api';

const getTeamLeaders = async () => {
  const response = await managerAPI.get('/team/team-leaders');
  return response.data;
};

const getTeamsByProject = async (pid) => {
  const response = await managerAPI.get(`/team/project/${pid}`);
  return response.data;
};

const createTeam = async (pid, domainId, tlUid) => {
  const response = await managerAPI.post('/team/create', { pid, domainId, tlUid });
  return response.data;
};

const addMember = async (pid, uid) => {
  const response = await managerAPI.post(`/team/add?pid=${pid}&uid=${uid}`);
  return response.data;
};

const removeMember = async (pid, uid) => {
  await managerAPI.delete(`/team/remove?pid=${pid}&uid=${uid}`);
};

const deleteTeam = async (teamId) => {
  await managerAPI.delete(`/team/${teamId}`);
};

export default {
  getTeamLeaders,
  getTeamsByProject,
  createTeam,
  addMember,
  removeMember,
  deleteTeam,
};
